### PR TITLE
Close #1484: adds rate tests for all tiers

### LINKED ISSUE #1484: Add tests for Rate stats (standalone and clustered)

### PR DESCRIPTION
**Pull request by gibsong**
Unit tests for statistics rate tests on all 4 tiers
1. heap
2. offheap
3. disk
4. clustered

task: https://github.com/ehcache/ehcache3/issues/1484

### CODE REVIEW

**Comment by gibsong** [comment]
test this please

**Comment by gibsong** [comment]
test this please

File: management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      //Put values in cache
+      cache.put(1L, "one");
+      cache.put(2L, "two");
+      cache.put(3L, "three");
+
+      cache.get(1L);//HIT lowest tier
+      cache.get(2L);//HIT lowest tier
+      cache.get(3L);//HIT lowest tier
+
+      cache.get(1L);//HIT higher tier
```
Comment (ljacomet) [inline management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java]: No, the comment assumes _again_ predictable eviction. See #1629

File: management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      cache.put(2L, "two");
+      cache.put(3L, "three");
+
+      cache.get(4L);//MISS
+      cache.get(5L);//MISS
+      cache.get(6L);//MISS
+
+      //TIER stats
+      for (int i = 0; i < statNames.size(); i++) {
+        double tierMissRate = StatsUtil.getExpectedValueFromRateHistory(statNames.get(i), context, managementRegistry, tierExpectedValues.get(i));
```
Comment (ljacomet) [inline management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java]: Please have a look at the clean up I did in #1629, you end up asserting twice the same thing for no added value

File: management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        Assert.assertThat(tierMissRate, is(tierExpectedValues.get(i)));
+      }
+
+      //CACHE stats
+      double cacheMissRate = StatsUtil.getExpectedValueFromRateHistory("Cache:MissRate", context, managementRegistry, CACHE_MISS_RATE);
```
Comment (ljacomet) [inline management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java]: same here - see cleanup in #1629

File: management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java (hunk 1 above)
Comment (gibsong) [inline management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java]: Shoot I forgot to add the eviction advisor.  I added it and tested it locally.

File: management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java (hunk 2 above)
Comment (gibsong) [inline management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java]: yeah that makes more sense.  I got rid of the unnecessary asserts.

**Comment by gibsong** [comment]
I added the task # to the commit message and I set my git user.name to "geoff gibson"  and user.email to "geoff.gibson@sofwareag.com"  I thought I had done this before, but now I made sure to set them globally. Thus you should only see this user.name and user.email from now on.

**Comment by gibsong** [comment]
test this please

**Comment by ljacomet** [comment]
@gibsong Your first commit still has different committer / author. Please fix that one as well.

**Comment by gibsong** [comment]
@ljacomet I updated the first commit to have the correct author.  Do you want me to go back in history and change all my commits to Author: geoff gibson <geoff.gibson@sofwareag.com>

**Comment by ljacomet** [comment]
@gibsong No was talking about these commits right here. We can't go back and modify others, that would require a force push on master which is a no-go.
Looks good now although I am going to leave the blocking _changes requested_ until the other stats test issues are fixed.

**Comment by gibsong** [comment]
test this please

**Comment by gibsong** [comment]
Tests passed!  please merge.

**Comment by ljacomet** [comment]
Good to go once rebased

File: management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+@RunWith(Parameterized.class)
+public class HitRateTest {
+
+  @Rule
+  public final Timeout globalTimeout = Timeout.seconds(10);
+
+  @Rule
+  public final TemporaryFolder diskPath = new TemporaryFolder();
+
+  private static final EhcacheStatisticsProviderConfiguration EHCACHE_STATISTICS_PROVIDER_CONFIG = new EhcacheStatisticsProviderConfiguration(1,TimeUnit.MINUTES,100,1,TimeUnit.MILLISECONDS,10,TimeUnit.MINUTES);
```
Comment (mathieucarbou) [inline management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java]: I would change the frequency to `1,TimeUnit.SECONDS`. Computing ALL stats each 1 millisecond is too heavy I think and will fill the circular buffer quite fast. You probably don't need such high frequency ?

Also, several stat tests have the same configuration. I would set back to 1 second, as @lorban  already suggested.

File: management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+@RunWith(Parameterized.class)
+public class MissRateTest {
+
+  @Rule
+  public final Timeout globalTimeout = Timeout.seconds(10);
+
+  @Rule
+  public final TemporaryFolder diskPath = new TemporaryFolder();
+
+  private static final EhcacheStatisticsProviderConfiguration EHCACHE_STATISTICS_PROVIDER_CONFIG = new EhcacheStatisticsProviderConfiguration(1,TimeUnit.MINUTES,100,1,TimeUnit.MILLISECONDS,10,TimeUnit.MINUTES);
```
Comment (mathieucarbou) [inline management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java]: same here

File: management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+  @Test
+  public void test() throws InterruptedException, IOException {
+
+    CacheManager cacheManager = null;
+
+    try {
+
+      DefaultManagementRegistryConfiguration registryConfiguration = new DefaultManagementRegistryConfiguration().setCacheManagerAlias("myCacheManager");
+      registryConfiguration.addConfiguration(new EhcacheStatisticsProviderConfiguration(1,TimeUnit.MINUTES,100,1,TimeUnit.MILLISECONDS,10,TimeUnit.MINUTES));
```
Comment (mathieucarbou) [inline management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java]: same here

**Comment by mathieucarbou** [comment]
You still have 2 commits in your PR with the first one being a "cleanup". I suggest you squash both on a single commit `Closes #1484 adds miss and hit rate tests for standalone and clustered...`

File: management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java (hunk 4 above)
Comment (gibsong) [inline management/src/test/java/org/ehcache/management/providers/statistics/HitRateTest.java]: I will make the change. I changed it in the latency tests but I also planned to update the rest of the stats test in a separate task.

File: management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java (hunk 5 above)
Comment (gibsong) [inline management/src/test/java/org/ehcache/management/providers/statistics/MissRateTest.java]: done

**Comment by gibsong** [comment]
@mathieucarbou that makes sense, I squashed the 2 commits together.

**Comment by gibsong** [comment]
test this please

**Comment by gibsong** [comment]
test this please

**Comment by gibsong** [comment]
@ljacomet  I just rebased.  The build passed.  If everything looks good can u please merge?

**Comment by ljacomet** [comment]
Good to go once rebased

**Comment by gibsong** [comment]
@ljacomet This code was rebased today and all test pass.  Can u please merge?

**Comment by mathieucarbou** [comment]
yes! ok for me.

**Pull request merged**